File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        flag = False
        stack = []

        #hashmap
        my_dict = {")" : "(", "}" : "{", "]" : "["}

        for char in s:

            if char in my_dict.values():
                stack.append(char)
            elif char in my_dict: #If we its a closing bracket..
                if not stack or stack.pop() != my_dict[char]:
                    flag = True
                    return False
        
        return len(stack) == 0
        

        #Everytime we see an opeing bracket. We want to append...
        #Ok is the stack empty? or the value at the topf of the stack is the my_dict[char] -> value...

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(std::string s) {
        
        std::stack<char> stack;

        for (int i = 0; i < s.size(); i++)
        {
            if (!stack.empty() && (s[i] == '}' || s[i] == ')' || s[i] == ']'))
            {
                char c = stack.top();
                
                switch (s[i])  // switch on the current closing bracket
                {
                    case ')':
                        if (c != '(') return false;
                        break;
                    case '}':
                        if (c != '{') return false;
                        break;
                    case ']':
                        if (c != '[') return false;
                        break;
                }

                stack.pop();
            }
            else
            {
                stack.push(s[i]);
            }
        }

        return (stack.size() == 0);
    }
};